### ISSUE TITLE
DFT trains the qwen3 model, the loss is quickly 0 and the gradient is very small

### ISSUE THREAD

**Issue by Yaawenn** [yongliang-wu/DFT#15]
I trained qwen3-0.6b on Llama-factory as follows:
```python
def dft_loss_func(outputs, labels, num_items_in_batch=None):
    logits = outputs.get("logits")
    if logits is None:
        return outputs.get("loss", torch.tensor(0.0))

    logits = logits.float()
    vocab_size = logits.size(-1)
    
    labels = torch.nn.functional.pad(labels, (0, 1), value=-100)
    shift_labels = labels[..., 1:].contiguous()
    logits = logits.view(-1, vocab_size)
    shift_labels = shift_labels.view(-1)
    shift_labels = shift_labels.to(logits.device)

    loss = _dft_cross_entropy(logits, shift_labels, num_items_in_batch)
    return loss


def _dft_cross_entropy(
    source: torch.Tensor,
    target: torch.Tensor,
    num_items_in_batch: Optional[torch.Tensor] = None,
    ignore_index: int = -100,
) -> torch.Tensor:
    valid_mask = target != ignore_index
    if not valid_mask.any():
        return torch.tensor(0.0, device=source.device, dtype=source.dtype)

    per_token_loss = torch.nn.functional.cross_entropy(source, target, ignore_index=ignore_index, reduction="none")
    valid_losses = per_token_loss[valid_mask]

    valid_source = source[valid_mask]
    valid_target = target[valid_mask]
    
    probs = torch.softmax(valid_source, dim=-1)
    valid_prob_coefficients = probs.gather(1, valid_target.unsqueeze(-1)).squeeze(-1)
    weighted_losses = valid_losses * valid_prob_coefficients.detach()


    if num_items_in_batch is not None:
        total_loss = weighted_losses.sum()
        if torch.is_tensor(num_items_in_batch):
            num_items_in_batch = num_items_in_batch.to(total_loss.device)
        loss = total_loss / num_items_in_batch
    else:
        loss = weighted_losses.mean()
    return loss
```

In the implementation of llama-factory, the chat template in tokenizer_config is automatically applied to the data. My example target is:
```
label_ids:
[-100, -100, -100, -100, ..., -100, -100, -100, -100, 151667, 271, 151668, 271, 51, 151645, 198]
labels:
<think>

</think>

T<|im_end|>
```

What is the reason for this phenomenon? Is it due to the end_token in the chat template?
Hope to get the answer, thank you!

**Comment by LUUUUUUZ** [comment]
I meet the same problem
